reposerver/repository/repository.go

```diff
@@ -1377,3 +1377,46 @@ func (s *Service) GetHelmCharts(ctx context.Context, q *apiclient.HelmChartsRequ
 	}
 	return &res, nil
 }
+
+func (s *Service) TestRepository(ctx context.Context, q *apiclient.TestRepositoryRequest) (*apiclient.TestRepositoryResponse, error) {
+	repo := q.Repo
+	checks := map[string]func() error{
+		"git": func() error {
+			return s.TestRepo(repo)
+		},
+		"helm": func() error {
+			if repo.EnableOCI {
+				_, err := helm.NewClient(repo.Repo, repo.GetHelmCreds(), repo.EnableOCI).TestHelmOCI()
+				return err
+			} else {
+				_, err := helm.NewClient(repo.Repo, repo.GetHelmCreds(), repo.EnableOCI).GetIndex(false)
+				return err
+			}
+		},
+	}
+	if check, ok := checks[repo.Type]; ok {
+		return &apiclient.TestRepositoryResponse{VerifiedRepository: false}, check()
+	}
+	var err error
+	for _, check := range checks {
+		err = check()
+		if err == nil {
+			return &apiclient.TestRepositoryResponse{VerifiedRepository: true}, nil
+		}
+	}
+	return &apiclient.TestRepositoryResponse{VerifiedRepository: false}, err
+}
+
+// TestRepo tests if a repo exists and is accessible with the given credentials
+func (s *Service) TestRepo(repo *v1alpha1.Repository) error {
+	clnt, err := s.newClient(repo)
```

Comment:
In `TestRepo`, you will need to add a repo lock. Why? The Git client (as implemented) use a single temporary directory (for example, `https:__github.com_argoproj-labs_applicationset`) for all operations on a repository.

However, what happens if multiple threads (goroutines) are both using that directory/repo at the same time? For example, what if one thread checks out branch 'my-feature' (causing the files on the disk to change to my-feature), and another thread, at the same time, checks out branch 'some-other-feature'? They will overwrite each other's files with these changes, producing undefined behaviour.

So in order to avoid this, there is the `repoLock.Lock` code, which intelligently manages this problem by using a lock and allowing concurrent operations in some cases.

You just need to add a call to repoLock, and it should work as expected. (The reason this wasn't required before was that some of the git client operations, like LsRefs, were in memory)

You can see an example of a repo lock being acquired here:
https://github.com/argoproj/argo-cd/blob/53fabec8455ccfbe9482c0a122fb091f2c38e25d/reposerver/repository/repository.go#L142

Reply:
Sure, I will make this change